umap/static/umap/js/modules/form/fields.js

```diff
@@ -0,0 +1,1366 @@
+    }
+    return opts
+  }
+}
+
+Fields.FacetSearchNumber = class extends Fields.MinMaxBase {
+  prepareForJS(value) {
+    return new Number(value)
+  }
+}
+
+Fields.FacetSearchDate = class extends Fields.MinMaxBase {
+  prepareForJS(value) {
+    return new Date(value)
+  }
+
+  toLocaleDateTime(dt) {
+    return new Date(dt.valueOf() - dt.getTimezoneOffset() * 60000)
+  }
+
+  prepareForHTML(value) {
+    // Value must be in local time
+    if (Number.isNaN(value)) return
+    return this.toLocaleDateTime(value).toISOString().substr(0, 10)
+  }
+
+  getLabels() {
+    return [translate('From'), translate('Until')]
+  }
+}
+
+Fields.FacetSearchDateTime = class extends Fields.FacetSearchDate {
+  getInputType(type) {
+    return 'datetime-local'
+  }
+
+  prepareForHTML(value) {
+    // Value must be in local time
+    if (Number.isNaN(value)) return
+    return this.toLocaleDateTime(value).toISOString().slice(0, -1)
+  }
+}
+
+Fields.MultiChoice = class extends BaseElement {
+  getDefault() {
+    return 'null'
+  }
+  // TODO: use public property when it's in our baseline
+  getClassName() {
+    return 'umap-multiplechoice'
+  }
+
+  clear() {
+    const checked = this.container.querySelector('input[type="radio"]:checked')
+    if (checked) checked.checked = false
+  }
+
+  fetch() {
+    this.initial = this.toHTML()
+    let value = this.initial
+    if (!this.container.querySelector(`input[type="radio"][value="${value}"]`)) {
+      value =
+        this.properties.default !== undefined ? this.properties.default : this.default
+    }
+    const choices = this.getChoices().map(([value, label]) => `${value}`)
+    if (choices.includes(`${value}`)) {
+      this.container.querySelector(`input[type="radio"][value="${value}"]`).checked =
+        true
+    }
+  }
+
+  value() {
+    const checked = this.container.querySelector('input[type="radio"]:checked')
+    if (checked) return checked.value
+  }
+
+  getChoices() {
+    return this.properties.choices || this.choices
+  }
+
+  getTemplate() {
+    return `<div class="${this.getClassName()} by${this.getChoices().length}" data-ref=wrapper></div>`
+  }
+
+  build() {
+    const choices = this.getChoices()
+    for (const [i, [value, label]] of choices.entries()) {
+      this.addChoice(value, label, i)
+    }
+    this.fetch()
+    super.build()
+  }
+
+  addChoice(value, label, counter) {
+    const id = `${Date.now()}.${this.name}.${counter}`
+    const input = Utils.loadTemplate(
+      `<input type="radio" name="${this.name}" id="${id}" value="${value}" />`
+    )
+    this.elements.wrapper.appendChild(input)
+    this.elements.wrapper.appendChild(
+      Utils.loadTemplate(`<label for="${id}">${label}</label>`)
+    )
+    input.addEventListener('change', () => this.sync())
+  }
+}
+
+Fields.TernaryChoices = class extends Fields.MultiChoice {
+  getDefault() {
+    return 'null'
+  }
+
+  toJS() {
+    let value = this.value()
+    switch (value) {
```

Comment:
Third time, can we move it to a dedicated util?

Reply:
Oops, good catch, NullableBoolean was duplicated (I guess we had one in FormBuilder and one in our custom fields, and I merged all in one file without noticing)! So I remove one, so we only have two switch remaining, and they are not exactly the same (handling `null` or not).